File: Week_02/java/src/_785_IsGraphBipartite.java
import java.util.LinkedList;

/**
 * 第一遍：2020/07/16周四 ✅
 * 第二遍：2020/07/16周四
 * 第三遍：2020/07/08周四
 * 第四遍：2020/07/13周一
 * 第三遍：2020/06/29周一
 * 第四遍：2020/07/05周日
 */
class _785_IsGraphBipartite {
    //2ms
    //时间复杂度O(N + M)N：表示点数 M：表示边数 空间O(N)  为什么时间复杂度要加上边数？
    public boolean isBipartite(int[][] graph) {
        if (graph == null || graph.length == 0) return false;
        int length = graph.length;
        int[] color = new int[length];
        LinkedList<Integer> queue = new LinkedList<>();
        for (int k = 0; k < length; k++) {
            if (color[k] != 0) continue;
            color[k] = 1;
            queue.offer(k);
            while (!queue.isEmpty()) {
                int i = queue.poll();
                for (int j : graph[i]) {
                    if (color[j] == 0) {
                        color[j] = -color[i];
                        queue.offer(j);
                    } else if (color[j] == color[i]) {
                        return false;
                    }
                }
            }
        }

        return true;
    }

    class Solution_DFS {
        //0ms
        public boolean isBipartite(int[][] graph) {
            if (graph == null || graph.length == 0) return false;
            int length = graph.length;
            int[] color = new int[length];
            for (int k = 0; k < length; k++) {
                if (color[k] != 0) continue;
                color[k] = 1;
                if (!dfs(graph, k, color)) return false;
            }

            return true;
        }

        boolean dfs(int[][] graph, int i, int[] color) {
            for (int j : graph[i]) {
                if (color[j] == 0) {
                    color[j] = -color[i];
                    if (!dfs(graph, j, color)) return false;
                } else if (color[j] == color[i]) {
                    return false;
                }
            }
            return true;
        }
    }
}

File: Week_02/dart/785_is_graph_bipartite.dart
import 'dart:collection';

void main(List<String> args) {
  print(Solution().isBipartite([
    [2, 2, 3],
    [0, 2],
    [0, 1, 3],
    [0, 2]
  ]));
  print(Solution().isBipartite([
    [1, 3],
    [0, 2],
    [1, 3],
    [0, 2]
  ]));

  print("-----------------");

  print(Solution2().isBipartite([
    [2, 2, 3],
    [0, 2],
    [0, 1, 3],
    [0, 2]
  ]));
  print(Solution2().isBipartite([
    [1, 3],
    [0, 2],
    [1, 3],
    [0, 2]
  ]));
}

class Solution2 {
  bool isBipartite(List<List<int>> graph) {
    final color = List.filled(graph.length, 0);
    for (int i = 0; i < graph.length; i++) {
      if (color[i] != 0) {
        continue;
      }
      color[i] = 1;
      if (!dfs(color, graph, i)) return false;
    }

    return true;
  }

  bool dfs(List<int> color, List<List<int>> graph, int i) {
    for (final j in graph[i]) {
      if (color[j] == 0) {
        color[j] = -color[i];
        if (!dfs(color, graph, j)) return false;
      } else if (color[j] == color[i]) {
        return false;
      }
    }
    return true;
  }
}

class Solution {
  bool isBipartite(List<List<int>> graph) {
    final color = List.filled(graph.length, 0);
    final q = Queue<int>();
    for (int i = 0; i < graph.length; i++) {
      if (color[i] != 0) {
        continue;
      }
      color[i] = 1;
      q.add(i);
      while (q.isNotEmpty) {
        final n = q.removeFirst();
        for (final j in graph[n]) {
          if (color[j] == 0) {
            q.add(j);
            color[j] = -color[n];
          } else if (color[j] == color[n]) {
            return false;
          }
        }
      }
    }

    return true;
  }
}
